external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutConfig.java

```diff
@@ -0,0 +1,298 @@
+    public static final int DEFAULT_MAX_UNCOMMITTED_OFFSETS = 10_000;    // 10,000 records
+
+    // Kafka property names
+    public interface Consumer {
+        String GROUP_ID = "group.id";
+        String BOOTSTRAP_SERVERS = "bootstrap.servers";
+        String ENABLE_AUTO_COMMIT = "enable.auto.commit";
+        String AUTO_COMMIT_INTERVAL_MS = "auto.commit.interval.ms";
+        String KEY_DESERIALIZER = "key.deserializer";
+        String VALUE_DESERIALIZER = "value.deserializer";
+    }
+
+    /**
+     * The offset used by the Kafka spout in the first poll to Kafka broker. The choice of this parameter will
+     * affect the number of consumer records returned in the first poll. By default this parameter is set to UNCOMMITTED_EARLIEST. <br/><br/>
+     * The allowed values are EARLIEST, LATEST, UNCOMMITTED_EARLIEST, UNCOMMITTED_LATEST. <br/>
+     * <ul>
+     * <li>EARLIEST means that the kafka spout polls records starting in the first offset of the partition, regardless of previous commits</li>
+     * <li>LATEST means that the kafka spout polls records with offsets greater than the last offset in the partition, regardless of previous commits</li>
+     * <li>UNCOMMITTED_EARLIEST means that the kafka spout polls records from the last committed offset, if any.
+     * If no offset has been committed, it behaves as EARLIEST.</li>
+     * <li>UNCOMMITTED_LATEST means that the kafka spout polls records from the last committed offset, if any.
+     * If no offset has been committed, it behaves as LATEST.</li>
+     * </ul>
+     * */
+    public enum FirstPollOffsetStrategy {
+        EARLIEST,
+        LATEST,
+        UNCOMMITTED_EARLIEST,
+        UNCOMMITTED_LATEST }
+
+    /**
+     * Defines when to poll the next batch of records from Kafka. The choice of this parameter will affect throughput and the memory
+     * footprint of the Kafka spout. The allowed values are STREAM and BATCH. STREAM will likely have higher throughput and use more memory
+     * (it stores in memory the entire KafkaRecord,including data). BATCH will likely have less throughput but also use less memory.
+     * The BATCH behavior is similar to the behavior of the previous Kafka Spout. De default value is STREAM.
+     * <ul>
+     *     <li>STREAM Every periodic call to nextTuple polls a new batch of records from Kafka as long as the maxUncommittedOffsets
+     *     threshold has not yet been reached. When the threshold his reached, no more records are polled until enough offsets have been
+     *     committed, such that the number of pending offsets is less than maxUncommittedOffsets. See {@link Builder#setMaxUncommittedOffsets(int)}
+     *     </li>
+     *     <li>BATCH Only polls a new batch of records from kafka once all the records that came in the previous poll have been acked.</li>
+     * </ul>
+     */
+    public enum PollStrategy {
+        STREAM,
+        BATCH
```

Comment:
Maybe I have misread the code for the original kafka spout, but I don't think that either of these even come close to doing what the original spout did so at a minimum we need to change the explanation.  What is the point of having this config at all?  BATCH is equivalent of setting maxUncommittedOffsets to 0, unless we decide to make it work more like the original spout.

For clarification the original `KafkaSpout` calls into the `PartitionManager` to emit a tuple.  It keeps a queue of messages that need to be emitted, but have not been emitted yet.  If that buffer were empty it attempts to refill the buffer with another batch from kafka.  Then it will pop off the first message try to transform it into a tuple and emit that tuple.  Only if the transformation turns into an empty tuple or something like that does it try again with another tuple form the queue.

https://github.com/apache/storm/blob/master/external/storm-kafka/src/jvm/org/apache/storm/kafka/PartitionManager.java#L136-L173

https://github.com/apache/storm/blob/master/external/storm-kafka/src/jvm/org/apache/storm/kafka/KafkaSpout.java#L139-L141

So the spout will only emit at most one event from kafka per call to nextTuple ever.  From looking at the code it looks like there is a bug in there if we do try to transform one event into multiple tuples where the tracking gets off.